algebra/group/coset.lean

```diff
@@ -0,0 +1,174 @@
+open coset_notation
+variable [has_mul G]
+
+lemma mem_lcoset {S : set G} {x : G} (a : G) (hxS : x ∈ S) : a * x ∈ a * S :=
+mem_image_of_mem (λ b : G, a * b) hxS
+
+lemma mem_rcoset {S : set G} {x : G} (a : G) (hxS : x ∈ S) : x * a ∈ S * a :=
+mem_image_of_mem (λ b : G, b * a) hxS
+
+def lcoset_equiv (S : set G) (a b : G) := a * S = b * S
+
+lemma lcoset_equiv_rel (S : set G) : equivalence (lcoset_equiv S) := 
+    mk_equivalence (lcoset_equiv S) (λ a, rfl) (λ a b, eq.symm) (λ a b c, eq.trans)
+
+end coset_mul
+
+section coset_semigroup
+open coset_notation
+variable [semigroup G]
+
+lemma lcoset_assoc (S : set G) (a b : G) : a * (b * S) = (a * b) * S :=
+    have h : (λ x : G, a * x) ∘ (λ x : G, b * x) = (λ x : G, (a * b) * x), from funext (λ c : G, 
+    calc
+        ((λ x : G, a * x) ∘ (λ x : G, b * x)) c = a * (b * c)               : by simp
+        ...                                     = (λ x : G, (a * b) * x) c  : by rw ← mul_assoc; simp ),
+    calc
+        a * (b * S) = image ((λ x : G, a * x) ∘ (λ x : G, b * x)) S         : by rw [lcoset, lcoset, ←image_comp]
+        ...         = (a * b) * S                                           : by rw [lcoset, h]
+
+lemma rcoset_assoc (S : set G) (a b : G) : S * a * b = S * (a * b) :=
+    have h : (λ x : G, x * b) ∘ (λ x : G, x * a) = (λ x : G, x * (a * b)), from funext (λ c : G, 
+    calc
+        ((λ x : G, x * b) ∘ (λ x : G, x * a)) c = c * a * b                 : by simp
+        ...                                     = (λ x : G, x * (a * b)) c  : by rw mul_assoc; simp),
+    calc
+        S * a * b   = image ((λ x : G, x * b) ∘ (λ x : G, x * a)) S         : by rw [rcoset, rcoset, ←image_comp]
+        ...         = S * (a * b)                                           : by rw [rcoset, h]
+
+lemma lcoset_rcoset (S : set G) (a b : G) : a * S * b = a * (S * b) := 
+    have h : (λ x : G, x * b) ∘ (λ x : G, a * x) = (λ x : G, a * (x * b)), from funext (λ c : G, 
+    calc
+        ((λ x : G, x * b) ∘ (λ x : G, a * x)) c = a * c * b                 : by simp
+        ...                                     = (λ x : G, a * (c * b)) c  : by rw mul_assoc; simp),
+    calc
+        a * S * b   = image ((λ x : G, x * b) ∘ (λ x : G, a * x)) S         : by rw [rcoset, lcoset, ←image_comp]
+        ...         = a * (S * b)                                           : by rw [rcoset, lcoset, ←image_comp, h]
+
+end coset_semigroup
+
+section coset_subgroup
+open subgroup coset_notation
+variables [group G] (S : set G) [subgroup S]
+
+lemma lcoset_mem_lcoset {a : G} (ha : a ∈ S) : a * S = S := 
+begin
+    simp [lcoset, image],
+    simp [set_eq_def, mem_set_of_eq],
+    intro b,
+    split,
+    { 
+        intro h,
+        cases h with x hx,
+        cases hx with hxl hxr,
+        rw [←hxr],
+        exact mul_mem ha hxl 
+    },
+    {   
+        intro h,
+        existsi a⁻¹ * b,
+        split,
+        have : a⁻¹ ∈ S, from inv_mem ha,
+        exact mul_mem this h,
+        simp
+    }
+end
```

Comment:
style: 2 indent spaces, braces on same line

Reply:
Fixed in https://github.com/leanprover/mathlib/pull/94/commits/91693ec5a4d0762766254c8a4d3153659349e094.